File: sdk/dotnet/Heartlight.Hip/SynapticKernel.cs
namespace Heartlight.Hip;

public sealed record SynapticConfig(
    double LearningRate = 0.08,
    double Decay = 0.002,
    double TraceDecay = 0.9,
    double MinWeight = -1.0,
    double MaxWeight = 1.0
);

public readonly record struct SynapticState(double Weight, double Eligibility);
public readonly record struct SynapticInput(double Pre, double Post, double Reward = 1.0);

public static class SynapticKernel
{
    public const string Version = "1.0";

    public static SynapticState Step(SynapticState state, SynapticInput input, SynapticConfig? config = null)
    {
        config ??= new SynapticConfig();
        double[] values =
        [
            config.LearningRate, config.Decay, config.TraceDecay, config.MinWeight, config.MaxWeight,
            state.Weight, state.Eligibility, input.Pre, input.Post, input.Reward
        ];
        if (values.Any(v => !double.IsFinite(v)))
            throw new ArgumentOutOfRangeException(nameof(input), "HEARTLIGHT synaptic values must be finite");
        if (config.LearningRate < 0 || config.Decay is < 0 or > 1 || config.TraceDecay is < 0 or > 1 || config.MinWeight > config.MaxWeight)
            throw new ArgumentOutOfRangeException(nameof(config));

        var eligibility = config.TraceDecay * state.Eligibility + input.Pre * input.Post;
        var rawWeight = (1.0 - config.Decay) * state.Weight + config.LearningRate * input.Reward * eligibility;
        var weight = Math.Clamp(rawWeight, config.MinWeight, config.MaxWeight);
        return new SynapticState(weight, eligibility);
    }
}

File: sdk/java/src/main/java/world/navis/heartlight/SynapticKernel.java
package world.navis.heartlight;

public final class SynapticKernel {
    public static final String VERSION = "1.0";

    public record Config(double learningRate, double decay, double traceDecay, double minWeight, double maxWeight) {
        public Config() { this(0.08, 0.002, 0.9, -1.0, 1.0); }
    }

    public record State(double weight, double eligibility) {}
    public record Input(double pre, double post, double reward) {}

    private SynapticKernel() {}

    public static State step(State state, Input input) {
        return step(state, input, new Config());
    }

    public static State step(State state, Input input, Config config) {
        double[] values = {
            config.learningRate(), config.decay(), config.traceDecay(), config.minWeight(), config.maxWeight(),
            state.weight(), state.eligibility(), input.pre(), input.post(), input.reward()
        };
        for (double value : values) {
            if (!Double.isFinite(value)) throw new IllegalArgumentException("HEARTLIGHT synaptic values must be finite");
        }
        if (config.learningRate() < 0 || config.decay() < 0 || config.decay() > 1 ||
            config.traceDecay() < 0 || config.traceDecay() > 1 || config.minWeight() > config.maxWeight()) {
            throw new IllegalArgumentException("invalid HEARTLIGHT synaptic config");
        }
        double eligibility = config.traceDecay() * state.eligibility() + input.pre() * input.post();
        double rawWeight = (1.0 - config.decay()) * state.weight() + config.learningRate() * input.reward() * eligibility;
        double weight = Math.max(config.minWeight(), Math.min(config.maxWeight(), rawWeight));
        return new State(weight, eligibility);
    }
}
